huaweicloud/services/dws/resource_huaweicloud_dws_logical_cluster.go

```diff
@@ -0,0 +1,503 @@
+	if err != nil || id == nil {
+		return diag.Errorf("error creating DWS logical cluster: ID is not found in API response")
+	}
+	d.SetId(id.(string))
+
+	return resourceLogicalClusterRead(ctx, d, meta)
+}
+
+func buildCreateLogicalClusterBodyParams(d *schema.ResourceData) map[string]interface{} {
+	return map[string]interface{}{
+		"logical_cluster": map[string]interface{}{
+			"logical_cluster_name": d.Get("logical_cluster_name"),
+			"cluster_rings":        buildLogicalClusterRingsRequestBody(d.Get("cluster_rings")),
+		},
+	}
+}
+
+func buildLogicalClusterRingsRequestBody(rawParams interface{}) []map[string]interface{} {
+	if rawSet, ok := rawParams.(*schema.Set); ok && rawSet.Len() > 0 {
+		rst := make([]map[string]interface{}, 0, rawSet.Len())
+		for _, v := range rawSet.List() {
+			raw, isMap := v.(map[string]interface{})
+			if !isMap {
+				continue
+			}
+
+			rst = append(rst, map[string]interface{}{
+				"ring_hosts": buildLogicalRingHostsRequestBody(raw["ring_hosts"]),
+			})
+		}
+		return rst
+	}
+	return nil
+}
+
+func buildLogicalRingHostsRequestBody(rawParams interface{}) []map[string]interface{} {
+	if rawSet, ok := rawParams.(*schema.Set); ok && rawSet.Len() > 0 {
+		rst := make([]map[string]interface{}, 0, rawSet.Len())
+		for _, v := range rawSet.List() {
+			raw, isMap := v.(map[string]interface{})
+			if !isMap {
+				continue
+			}
+
+			rst = append(rst, map[string]interface{}{
+				"host_name": raw["host_name"],
+				"back_ip":   raw["back_ip"],
+				"cpu_cores": raw["cpu_cores"],
+				"memory":    raw["memory"],
+				"disk_size": raw["disk_size"],
+			})
+		}
+		return rst
+	}
+	return nil
+}
+
+func waitingForStateCompleted(ctx context.Context, client *golangsdk.ServiceClient, d *schema.ResourceData,
+	timeout time.Duration) error {
+	clusterName := d.Get("logical_cluster_name").(string)
+	expression := fmt.Sprintf("logical_clusters[?logical_cluster_name=='%s']|[0]", clusterName)
+	stateConf := &resource.StateChangeConf{
+		Pending: []string{"PENDING"},
+		Target:  []string{"COMPLETED"},
+		Refresh: func() (interface{}, string, error) {
+			clusterRespBody, err := readLogicalClusters(client, d)
+			if err != nil {
+				return nil, "ERROR", err
+			}
+
+			cluster := utils.PathSearch(expression, clusterRespBody, nil)
+			if cluster == nil {
+				return nil, "ERROR", golangsdk.ErrDefault404{}
+			}
+
+			completed := utils.PathSearch("action_info.completed", cluster, false).(bool)
+			result := utils.PathSearch("action_info.result", cluster, "").(string)
+			if completed && result == "success" {
+				return cluster, "COMPLETED", nil
+			}
+
+			if completed && result == "failed" {
+				return cluster, "ERROR", fmt.Errorf("the DWS logical cluster (%s) is failed", clusterName)
+			}
+			return cluster, "PENDING", nil
+		},
+		Timeout:      timeout,
+		Delay:        30 * time.Second,
+		PollInterval: 30 * time.Second,
+	}
+	_, err := stateConf.WaitForStateContext(ctx)
```

Comment:
it is better to return the result of this func, so it is unnecessary to get the cluster info again in the func of resourceLogicalClusterCreate

Reply:
OK